### ISSUE TITLE
Jetty client: workaround for 'max requests per destination exceeded'

### ISSUE THREAD

**Issue by dmitrymurashenkov** [eclipse/jetty.project#862]
I've been using jetty client to perform basic load testing and run into this issue. Case is simple - I want to test max request per second with several threads/connections and server takes about 100ms to respond to each request

I perform `httpClient.send(CompletionListener)` in for loop and quickly hit max requests per destination since server is slow.

I've made workaround in my code: created blocking queue with size equal to max requests per destination. Before sending request I perform `queue.offer(requestId, 1, SECONDS)` and after receiving an answer I do `queue.remove(requestId)`. This helps me wait till I can safely send more requests.

Somehow it would be nice to have this feature in jetty client. For example:

``` java
boolean httpClient.send(CompletionListener, long timeout, TimeUnit unit)
```

or some better alternative which allows avoiding exceptions.

**Comment by sbordet** [comment]
The exception you get can be avoided by configuring a large number of requests per destination via `HttpClient.setMaxRequestsQueuedPerDestination()`.

Have you tried that ?

**Comment by dmitrymurashenkov** [comment]
Yes, I am aware of this property, but it seems not acceptable in my case.

I "generate" requests in for loop faster than they are processed, so even if jetty client doesn't reject them still they will eventually fill the heap and cause OOM. So I need to wait somewhere before generating next request.

**Comment by sbordet** [comment]
Then it seems appropriate that if you generate in such a way, then _you_ also control backpressure - not `HttpClient`.

What would your proposed `send(CompletionListener listener, long timeout, TimeUnit unit)` do ? Block ? And if the timeout expires, what ? Fail the request ?

The solution seems brittle, since you change to faster hardware and you have to change the timeouts, but at that point you can also change the max queue size. Having `HttpClient` rejecting the request because of max queue size exceeded or because of a timeout elapsed does not seem much different to me.

I feel this enhancement is very specific to the way you have coded your load test, so I am reluctant to put this into `HttpClient`.

If you have further insights that are more generic we can consider them, but so far I'm hesitant to consider your proposed changes for addition.

Thanks !

**Comment by dmitrymurashenkov** [comment]
Yeah, in my particular case numerous other workarounds are possible, so it is not critical. But still let me provide some explanations in case you consider some enhancement in future.

My initial thought was that it would be nice to provide API similar to `java.util.concurrent` classes to allow caller to handle this case in some way. 

Option 1 is `BlockingQueue` style API:

`boolean send(CompletionListener listener, long timeout, TimeUnit unit)`

This tries to send request and waits for space in queue to become available, returns true if succeeded, false otherwise (request was not sent).

Option 2 is `Executor` style API:

`void send(CompletionListener) throws RejectedExecutionException`
`setRejectedExecutionHandler(RejectedExecutionHandler)`

This either accepts task or throws exception. Perhaps even `CallerRunsPolicy` can be used here also in some way.

In real cases this problem is not likely to happen often, but still let's take a look at more realistic example and check pros and cons of current and proposed approach. 

Suppose we are writing some system that accepts input requests and performs an outgoing http request via jetty client for each input request. For example a proxy server or an app that uses social networks http API.

Jetty client queue can become full in 2 main cases: permanent overload or spikes. 

**Permanent overload**

Obviously we will have to discard some requests to bring actual request rate to some target value. One way to do it is indeed by changing jetty client queue size. However target rate depends on hardware, cpu/memory consumption by other running threads/processes, health of server we send requests to and can vary other time.  Limiting queue size stays constant and can't adjust itself to current conditions. 

Also it wastes some CPU time. Consider following example: we start a stress test to find out how our proxy app will behave if overloaded. We send too many input requests, each request requires some "preparation" work before we pass it to jetty client. Typical case would be:
we have some thread pool (either thread from http server pool invokes our servlet or we have our own intermediate pool which processes previously accepted requests) which does some request verifications/transformations, then passes request to jetty client and there request gets rejected because queue is full. 

If we could block until space in queue becomes available then our intermediate thread pool's threads would be idle and won't do redundant request processing until we are sure that there is space in queue. Of course in this case queue in our intermediate thread pool will become full too, but that just means that we will reject input requests earlier without doing processing. I guess this can be achieved via other means with existing API, but blocking API should make this easier.

**Spiking load**

In case of short spikes our whole proxy system can reject less requests if it uses blocking API. 

For example we have an intermediate pool with 1 thread (that invoke jetty client API) and queue size of 1 and jetty client with queue size of 1 and 1 connection per host. So intermediate thread pool can have 1 task beeing processed by thread and 1 task in queue and jetty can have 1 task in queue. 

Now 3 requests arrive. If our intermediate pool processing is fast then they will all will arrive to jetty client one by one but with little delay, 1 would be sent and 2 would be rejected by jetty's queue.

With blocking API 1 request would be sent, 1 would be blocked on jetty and 1 would pile up in our intermediate pool's queue. Profit! :)

**Comment by sbordet** [comment]
I'm going to close this as I believe that https://github.com/jetty-project/jetty-reactive-httpclient/ is a better solution to control backpressure.
